oauth_tools/external_idp.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import abc
+import logging
+from os.path import join
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, KubeConfig, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Namespace, Pod, Service
+from requests.exceptions import RequestException
+
+from oauth_tools.constants import (
+    DEX_CLIENT_ID,
+    DEX_CLIENT_SECRET,
+    DEX_MANIFESTS,
+    EXTERNAL_USER_EMAIL,
+    EXTERNAL_USER_PASSWORD,
+    KUBECONFIG,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalIdpService(abc.ABC):
```

Comment:
Should this also have a `login` method?

It could be take a playwright page as input and perform the login flow. This would allow us to further abstract `complete_auth_code_login` and `complete_device_login`.